### PR TITLE
Refactor arena synchronization

### PR DESCRIPTION
**Pull request by jasone**
[#621 redux]

This series of changes refactors out arena->lock, through a combination of atomic operations (especially for stats) and field-specific independent mutexes.

This relates to #466.

### CODE REVIEW

File: include/jemalloc/internal/prof_inlines_a.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+JEMALLOC_INLINE bool
+prof_accum_add(tsdn_t *tsdn, prof_accum_t *prof_accum, uint64_t accumbytes) {
+	cassert(config_prof);
+
+	bool overflow;
+	uint64_t a0, a1;
+
+	/*
+	 * If the application allocates fast enough (and/or if idump is slow
+	 * enough), extreme overflow here (a1 >= prof_interval * 2) can cause
```
Comment (interwq) [inline include/jemalloc/internal/prof_inlines_a.h]: Probably mention the case of very small prof_interval? Or that's what you mean by slow idump?

File: include/jemalloc/internal/prof_inlines_a.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	 * idump trigger coalescing.  This is an intentional mechanism that
+	 * avoids rate-limiting allocation.
+	 */
+#ifdef JEMALLOC_ATOMIC_U64
+	do {
+		a0 = atomic_read_u64(&prof_accum->accumbytes);
+		a1 = a0 + accumbytes;
```
Comment (interwq) [inline include/jemalloc/internal/prof_inlines_a.h]: nit: add assert(a1 >= a0) to make sure no uint64 overflow.

File: include/jemalloc/internal/prof_inlines_a.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	 * possible without underflowing.  Interval-triggered dumps occur
+	 * slightly more often than intended as a result of incomplete
+	 * canceling.
+	 */
+	uint64_t a0, a1;
+#ifdef JEMALLOC_ATOMIC_U64
+	do {
+		a0 = atomic_read_u64(&prof_accum->accumbytes);
+		a1 = (a0 >= LARGE_MINCLASS - usize) ?  a0 - LARGE_MINCLASS -
+		    usize : 0;
```
Comment (interwq) [inline include/jemalloc/internal/prof_inlines_a.h]: Should be ... ? a0 - (LARGE_MINCLASS - usize) : 0

File: include/jemalloc/internal/prof_inlines_a.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	} while (atomic_cas_u64(&prof_accum->accumbytes, a0, a1));
+#else
+	malloc_mutex_lock(tsdn, &prof_accum->mtx);
+	a0 = prof_accum->accumbytes;
+	a1 = (a0 >= LARGE_MINCLASS - usize) ?  a0 - LARGE_MINCLASS - usize : 0;
```
Comment (interwq) [inline include/jemalloc/internal/prof_inlines_a.h]: Here as well.

File: src/arena.c [hunk 5]
```diff
@@ -37,6 +37,207 @@ static void	arena_bin_lower_slab(tsdn_t *tsdn, arena_t *arena,
...
+	malloc_mutex_assert_owner(tsdn, &arena_stats->mtx);
+	*p += x;
+#endif
+}
+
+static void
+arena_stats_sub_u64(tsdn_t *tsdn, arena_stats_t *arena_stats, uint64_t *p,
+    uint64_t x) {
+#ifdef JEMALLOC_ATOMIC_U64
+	atomic_sub_u64(p, x);
```
Comment (interwq) [inline src/arena.c]: nit: add assertion to verify no underflow before sub

File: src/arena.c [hunk 6]
```diff
@@ -37,6 +37,207 @@ static void	arena_bin_lower_slab(tsdn_t *tsdn, arena_t *arena,
...
+
+	arena_stats_unlock(tsdn, &arena->stats);
+
+	if (config_tcache) {
+		tcache_bin_t *tbin;
+		tcache_t *tcache;
+
+		/* tcache_bytes counts currently cached bytes. */
+		astats->tcache_bytes = 0;
+		ql_foreach(tcache, &arena->tcache_ql, link) {
```
Comment (interwq) [inline src/arena.c]: This part needs to be protected as we walk through the tcache_ql (used to be under arena lock).

File: src/arena.c [hunk 7]
```diff
@@ -177,52 +364,36 @@ arena_large_dalloc_stats_update(arena_t *arena, size_t usize) {
...
+arena_large_reset_stats_cancel(tsdn_t *tsdn, arena_t *arena, size_t usize) {
 	szind_t index = size2index(usize);
 	szind_t hindex = (index >= NBINS) ? index - NBINS : 0;
 
 	cassert(config_stats);
 
-	arena->stats.ndalloc_large--;
-	arena->stats.lstats[hindex].ndalloc--;
+	arena_stats_lock(tsdn, &arena->stats);
+	arena_stats_sub_u64(tsdn, &arena->stats, &arena->stats.ndalloc_large,
```
Comment (interwq) [inline src/arena.c]: It's unclear to me that why this section isn't cancelling the same changes as in arena_large_dalloc_stats_update().

File: src/large.c [hunk 8]
```diff
@@ -286,20 +286,26 @@ large_ralloc(tsdn_t *tsdn, arena_t *arena, extent_t *extent, size_t usize,
...
  * junked_locked indicates whether the extent's data have been junk-filled, and
- * whether the arena's lock is currently held.  The arena's large_mtx is
- * independent of these considerations.
+ * whether the arena's large_mtx is currently held.
  */
 static void
 large_dalloc_prep_impl(tsdn_t *tsdn, arena_t *arena, extent_t *extent,
     bool junked_locked) {
-	malloc_mutex_lock(tsdn, &arena->large_mtx);
+	if (!junked_locked) {
```
Comment (interwq) [inline src/large.c]: nit: maybe it looks a bit more clear merging the branches into a single one:
```
if (!junked_locked) {
  malloc_mutex_lock(tsdn, &arena->large_mtx);
  extent_list_remove(&arena->large, extent);
  malloc_mutex_unlock(tsdn, &arena->large_mtx);
  large_dalloc_maybe_junk(...);
} else {
  malloc_mutex_assert_owner(tsdn, &arena->large_mtx);
  extent_list_remove(&arena->large, extent);
}
```

**Comment by interwq** [comment]
Finished reviewing the current version. Looking forward to testing this!

File: src/arena.c [hunk 9]
```diff
@@ -37,6 +37,207 @@ static void	arena_bin_lower_slab(tsdn_t *tsdn, arena_t *arena,
...
 
+static bool
+arena_stats_init(tsdn_t *tsdn, arena_stats_t *arena_stats) {
+#ifndef JEMALLOC_ATOMIC_U64
+	if (malloc_mutex_init(&arena_stats->mtx, "arena_stats",
+	    WITNESS_RANK_ARENA_STATS)) {
+		return true;
+	}
+#endif
+	/* Memory is zeroed, so there is no need to clear stats. */
```
Comment (jasone) [inline src/arena.c]: During the refactor I started writing assertions for all the fields, but there are so many fields the only reasonable approach is bytewise.  I wasn't sure of the utility, but will add checking.

File: include/jemalloc/internal/prof_inlines_a.h (hunk 3 above)
Comment (jasone) [inline include/jemalloc/internal/prof_inlines_a.h]: Good catch.

File: src/arena.c (hunk 6 above)
Comment (jasone) [inline src/arena.c]: My tree shows that f01d41f44efdf537f8c0046e3fe32c8eb83ce8eb indeed adds that locking.  Are you reviewing some mixture of the merged pull request and incremental versions of the changes?

Comment (interwq) [inline src/arena.c]: Yeah that's true. I was looking at them separately and forgot to discard that comment.

File: src/arena.c (hunk 5 above)
Comment (jasone) [inline src/arena.c]: There's actually an underflow for ```stats.mapped``` that I already have a fix for (based on this stack of diffs).  I'll add the assertions to that diff.

Comment (jasone) [inline src/arena.c]: For reference: [ephemeral link](https://github.com/jasone/jemalloc/commit/956f395e301476d159ebc001a93f8889c35c17cd)

**Pull request merged**